liquidprompt

```diff
@@ -4797,11 +4797,18 @@ _lp_default_theme_prompt() {
 # Construct the prompt #
 ########################
 
+_LP_NOT_A_REAL_COMMAND=0
 __lp_set_prompt() {
     # Display the return value of the last command, if different from zero
     # As this get the last returned code, it should be called first
     local -i lp_error="$?"
 
+    if (( _LP_NOT_A_REAL_COMMAND )); then
```

Comment:
Since it is possible for this to false flag on Bash (if the command that a user ran also existed in `PROMPT_COMMAND`), this should have an extra check. How about `&&` this with `$lp_error == 130`. There still is a chance for a false flag (if the command that a user ran also existed in `PROMPT_COMMAND`, then they `Ctrl^C`ed it), but it would be very unlikely.

Reply:
There seem to be another way to detect a real command: check if `$BASH_COMMAND` match `$(builtin builtin history 1)` ( `__lp_get_last_command_line()` but without the fallback) rather than _not_ match `$PROMPT_COMMAND`. What do you think?

Reply:
The problem is anything with `$( ... )` syntax creates a subshell, which will be a `fork()` system call. That is slow (relatively), and we want to avoid slowing down prompt generation as much as possible.

Reply:
Indeed. I was trying to figure something out but I am afraid checking `PROMPT_COMMAND` is the best we can get.

Regarding the `$lp_error == 130` condition — it isn't desired because of the following scenario:

1. Run a real command that returns an error other than 130 (e.g. `/bin/false`)
2. The error indicator appears.
3. Press [Enter]

Actual result:
The error indicator keeps appearing.

Expected result:
The error indicator disappears.

Reply:
I don't think that condition would happen if you `&&` it with `! _LP_REAL_COMMAND`. I will admit I did not test it.

Reply:
I am checking this at the moment and it works as I predicted. Upon entering `__lp_set_prompt()`:
1. `$?` (and `lp_error`) is 1
2. `_LP_REAL_COMMAND_` is 0
3. `LP_HIDE_EMPTY_ERROR` is 1

Thus, the condition `(( ! _LP_REAL_COMMAND && LP_HIDE_EMPTY_ERROR && $lp_error == 130 ))` evaluates to false and `lp_error` is still 1 and shows up at the prompt.

I am pushing a commit without the `lp_error` check. You can add it and see for yourself.

Reply:
Ah, I misread your previous comment. I see now what you mean.

That actually already happens. Run your test on the `master` branch and you will see that is nothing new. I think Bash and Zsh do not change the value of `$?` if you hit Enter on an empty prompt. If that is the case, there is not much we can do to fix it. And bonus, once this PR is merged, you can Ctrl^C to remove it if you want. If it is a problem in Liquid Prompt somehow, I'd be happy to fix it. But I'm pretty confident it's not; try this minimal PS1: `PS1='($?) \$ '`:

```
$ PS1='($?) \$ '
(0) $ false
(1) $
(1) $
```

So adding the ` && lp_error == 130` to the check is still the correct logic here.

Reply:
> That actually already happens

Yes and as far as I am concerned I don't like this behavior (I consider it invalid UI-wise). I don't assume I am right, this is just my opinion, but this is also what I'd like to change in this commit. My idea is: the error indicator is visible iff the last typed _real_ command returned an error. Neither of the following is a _real_ command (and they don't set `_LP_REAL_COMMAND`) in my understanding:

- an empty line
- a comment (a command starting with `#`, which also preserves an old value of `$?`)
- Ctrl-C

Therefore a prompt displayed after any of these events IMHO should not indicate an error.

> there is not much we can do to fix it

Yes. we can. Without the `&& $lp_error == 130` condition and with `_LP_REAL_COMMAND` set to 0 in all three above cases we can hide the _stale_ error code from users by setting `lp_error`  back to 0.

```sh
    local -i lp_error="$?"

    if (( ! _LP_REAL_COMMAND && LP_HIDE_EMPTY_ERROR )); then
        lp_error=0
    fi
```
At this point `$?` ma still be non-zero but we use `lp_error` to construct the prompt.

> And bonus, once this PR is merged, you can Ctrl^C to remove it if you want

That is because
1. ^C changes `$?` to 130
2. With `$lp_error == 130` check `lp_error` is set to 0 (but `$?` is still 130)

> If it is a problem in Liquid Prompt somehow, I'd be happy to fix it

No, it is not a problem in Liquid Prompt and thanks to `lp_error` being used to build prompt we can (by avoiding the `$lp_error == 130` condition) hide errors if they are not a result of a _real_ command.

But once again, this is my opinion. I am trying hard to explain it and back it up, but it is still an opinion. If you say we hide _only_ 130, I will add the condition.

Reply:
Ok, I think I understand you now. So you want two separate enhancements:
1. If [Enter] on empty or empty-with-comment line: clear previous error code.
2. If Ctrl^C on prompt (not on running command): hide error code.

They are related, obviously. And they are both valid. I think they are closely related enough that they can both go in the same PR.

The only remaining question is if both should be covered by the same config option. I think it is valid to do so, but the docs will need to be _very_ explicit in what scenarios it will hide the error code.

With all that in light, I'll do an updated review.